File: server/content/intake.js
// Working out what a piece of content IS, for free, and preparing it for a
// model to actually look at when — and only when — the user says what they
// want.
//
// The split between this file and investigator.js is the load-bearing part
// of the whole rebuild: `identify()` below makes NO model call, ever. It
// answers "what is this" (a title, a rough kind, a length) from cheap
// metadata alone — YouTube's oEmbed endpoint, a page's <title>, a file's
// size on disk. That is what lets Jarvis say "that's a 40-minute video on
// dropshipping — what do you want from it?" and then genuinely wait,
// instead of reading or watching the thing first and asking afterwards.
//
// A leaf module: imports adapters and fs/path, never the skills loader, the
// runner, or investigator.js (investigator.js imports this, not the other
// way around).

import fs from 'node:fs';
import path from 'node:path';
import { getAdapter, getCapabilities } from '../adapters/index.js';
import { fetchArticle, fileKind, mimeTypeFor, inlineAttachment } from '../media.js';

const BROWSER_UA =
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0 Safari/537.36';

// YouTube serves a cookie-consent interstitial instead of the real watch page
// to a request with no consent cookie. These are the standard "consent
// already given" cookies a browser would carry.
const YOUTUBE_HEADERS = {
  'User-Agent': BROWSER_UA,
  'Accept-Language': 'en-US,en;q=0.9',
  Cookie: 'CONSENT=YES+cb.20210328-17-p0.en+FX+000; SOCS=CAI',
};

const TRANSCRIPT_MAX_CHARS = 40000;

function decodeEntities(text) {
  return String(text || '')
    .replace(/&nbsp;/g, ' ')
    .replace(/&#(\d+);/g, (_, c) => String.fromCharCode(Number(c)))
    .replace(/&#x([0-9a-f]+);/gi, (_, c) => String.fromCharCode(parseInt(c, 16)))
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&amp;/g, '&');
}

// ---------- working out what the user gave us ----------

const YOUTUBE_HOSTS = /^(www\.|m\.|music\.)?(youtube\.com|youtu\.be)$/i;

/** The YouTube video id in a watch/short/youtu.be/embed URL, or null. */
export function youtubeVideoId(url) {
  try {
    const u = new URL(url);
    if (!YOUTUBE_HOSTS.test(u.hostname)) return null;
    if (u.hostname.toLowerCase().endsWith('youtu.be')) {
      return u.pathname.slice(1).split('/')[0] || null;
    }
    const v = u.searchParams.get('v');
    if (v) return v;
    const m = /^\/(shorts|embed|live|v)\/([^/?#]+)/.exec(u.pathname);
    return m ? m[2] : null;
  } catch {
    return null;
  }
}

/**
 * Works out what a piece of user input actually is. Deliberately forgiving —
 * fed by both a text box and a voice transcript, so it must cope with a
 * spoken path, a pasted link with tracking junk on it, or a wall of text
 * with no marker at all.
 *
 * Returns { kind: 'youtube'|'url'|'file'|'text', ... }. Purely syntactic —
 * no network call, no filesystem check.
 */
export function classifySource(input) {
  const raw = String(input || '').trim();
  if (!raw) return { kind: 'text', text: '' };

  const videoId = youtubeVideoId(raw);
  if (videoId) return { kind: 'youtube', url: raw, videoId };

  if (/^https?:\/\//i.test(raw)) {
    try {
      const u = new URL(raw);
      return { kind: 'url', url: u.toString() };
    } catch {
      // Malformed link — treat it as text rather than failing outright.
    }
  }

  // A Windows path (C:\... or \\server\share) or a quoted one, which is what
  // "Copy as path" puts on the clipboard.
  const unquoted = raw.replace(/^"(.*)"$/s, '$1').trim();
  if (/^[a-zA-Z]:[\\/]/.test(unquoted) || unquoted.startsWith('\\\\')) {
    return { kind: 'file', filePath: unquoted };
  }

  return { kind: 'text', text: raw };
}

function humanSize(bytes) {
  if (!bytes) return null;
  if (bytes < 1024 * 1024) return `${Math.round(bytes / 1024)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

// ---------- YouTube metadata without watching ----------

/** Title and channel from YouTube's public oEmbed endpoint — no key, no scraping. The cheapest possible "what is this" for a video. */
async function fetchYoutubeOEmbed(url) {
  try {
    const res = await fetch(
      `https://www.youtube.com/oembed?url=${encodeURIComponent(url)}&format=json`,
      { headers: { 'User-Agent': BROWSER_UA } }
    );
    if (!res.ok) return null;
    const data = await res.json();
    return { title: data.title || null, author: data.author_name || null, thumbnail: data.thumbnail_url || null };
  } catch {
    return null;
  }
}

/**
 * Extracts the JSON array that follows `"<key>":` in a blob of page source.
 * A regex can't do this — entries contain nested arrays, so a non-greedy
 * match truncates. Count brackets, respecting strings and escapes.
 */
function extractJsonArrayAfter(source, key) {
  const marker = `"${key}":`;
  const at = source.indexOf(marker);
  if (at === -1) return null;

  const start = source.indexOf('[', at);
  if (start === -1) return null;

  let depth = 0;
  let inString = false;
  let escaped = false;
  for (let i = start; i < source.length; i++) {
    const ch = source[i];
    if (escaped) {
      escaped = false;
      continue;
    }
    if (ch === '\\') {
      escaped = true;
      continue;
    }
    if (ch === '"') inString = !inString;
    if (inString) continue;
    if (ch === '[') depth++;
    else if (ch === ']') {
      depth--;
      if (depth === 0) return source.slice(start, i + 1);
    }
  }
  return null;
}

/** Pulls the caption text out of a timedtext track (XML `<text>` nodes). */
function parseCaptionXml(xml) {
  const lines = [];
  const re = /<text[^>]*>([\s\S]*?)<\/text>/g;
  let match;
  while ((match = re.exec(xml))) {
    const line = decodeEntities(match[1].replace(/<[^>]+>/g, ' ')).replace(/\s+/g, ' ').trim();
    if (line) lines.push(line);
  }
  return lines.join(' ');
}

/**
 * Everything about a YouTube video that can be had without a model watching
 * it: title, channel, description, and the caption transcript when one is
 * published. Only called when a video-capable model isn't available or
 * didn't work — this is the honest floor, not the default path.
 *
 * IN PRACTICE, EXPECT NO TRANSCRIPT — YouTube currently serves empty caption
 * bodies to anything without a session proof-of-origin token. Title and
 * description come from sturdier sources and do work. That's exactly why
 * watching properly matters when it's available: without it, this is what
 * Jarvis has to go on, and the caller must say so rather than imply more.
 */
export async function fetchYoutubeText(url) {
  const meta = (await fetchYoutubeOEmbed(url)) || {};
  const result = {
    ok: true,
    title: meta.title || null,
    author: meta.author || null,
    thumbnail: meta.thumbnail || null,
    description: null,
    transcript: null,
  };

  let html = '';
  try {
    const withLocale = url.includes('?') ? `${url}&hl=en&gl=US` : `${url}?hl=en&gl=US`;
    const res = await fetch(withLocale, { headers: YOUTUBE_HEADERS });
    if (res.ok) html = await res.text();
  } catch {
    // No watch page — oEmbed data alone still lets the caller say what the video is.
  }

  if (!html) {
    if (!result.title) return { ok: false, error: "Couldn't reach that video." };
    return result;
  }

  const desc = /"shortDescription":"((?:\\.|[^"\\])*)"/.exec(html);
  if (desc) {
    try {
      result.description = JSON.parse(`"${desc[1]}"`);
    } catch {
      result.description = desc[1];
    }
  }

  const tracksJson = extractJsonArrayAfter(html, 'captionTracks');
  if (tracksJson) {
    try {
      const tracks = JSON.parse(tracksJson.replace(/\\u0026/g, '&'));
      const track =
        tracks.find((t) => /^en/i.test(t.languageCode || '') && t.kind !== 'asr') ||
        tracks.find((t) => /^en/i.test(t.languageCode || '')) ||
        tracks[0];
      if (track?.baseUrl) {
        const capRes = await fetch(track.baseUrl, { headers: YOUTUBE_HEADERS });
        if (capRes.ok) {
          const transcript = parseCaptionXml(await capRes.text());
          if (transcript) result.transcript = transcript.slice(0, TRANSCRIPT_MAX_CHARS);
        }
      }
    } catch {
      // Captions unavailable — the caller falls back to title + description.
    }
  }

  return result;
}

// ---------- the free glance ----------

/**
 * Works out what a piece of content IS — no model call, ever. This is the
 * whole point of the split: sharing something must cost nothing until the
 * user says what they want done with it.
 *
 * For a URL, this does fetch the page (needed for even a title) and caches
 * the extracted text on the result as `cachedText` — one fetch serves both
 * "what is this" and, later, the actual reading, rather than fetching twice.
 */
export async function identify(source) {
  if (source.kind === 'youtube') {
    const meta = await fetchYoutubeOEmbed(source.url);
    return {
      title: meta?.title || `YouTube video (${source.videoId})`,
      kind: 'video',
      detail: meta?.author ? `on ${meta.author}` : null,
      thumbnail: meta?.thumbnail || null,
    };
  }

  if (source.kind === 'url') {
    const article = await fetchArticle(source.url);
    if (!article.ok) {
      return { title: source.url, kind: 'article', detail: null, unreachable: article.error };
    }
    return {
      title: article.title || source.url,
      kind: 'article',
      detail: `about ${Math.max(1, Math.round(article.text.length / 1000))}k characters`,
      cachedText: article.text,
    };
  }

  if (source.kind === 'file') {
    const kind = fileKind(source.filePath); // video | audio | image | document | unknown
    let size = null;
    try {
      size = fs.statSync(source.filePath).size;
    } catch {
      return { title: path.basename(source.filePath), kind, detail: null, unreachable: "I can't find that file." };
    }
    return { title: path.basename(source.filePath), kind, detail: humanSize(size) };
  }

  // Pasted/dictated text.
  const text = String(source.text || '');
  const firstLine = text.split(/\r?\n/)[0].trim();
  return {
    title: firstLine ? (firstLine.length > 60 ? `${firstLine.slice(0, 60)}…` : firstLine) : 'Pasted text',
    kind: 'text',
    detail: `${text.length} characters`,
    cachedText: text,
  };
}

/** Plain-language "what this is", for Jarvis to say out loud right after the free glance — before it has read or watched anything. */
export function describe(identity) {
  const bits = [];
  const kindWord = { video: 'a video', article: 'a web article', document: 'a document', image: 'an image', audio: 'an audio file', text: 'some pasted text', unknown: 'a file' };
  bits.push(kindWord[identity.kind] || 'something');
  if (identity.title) bits.push(`— "${identity.title}"`);
  if (identity.detail) bits.push(`(${identity.detail})`);
  return bits.join(' ');
}

/** Plain-language description of how content was actually taken in — said honestly, every time, so a verdict never sounds more informed than it is. */
export function intakeDescription(kind) {
  switch (kind) {
    case 'video':
      return 'I watched the video itself, including what was on screen.';
    case 'audio':
      return 'I listened to the audio.';
    case 'image':
      return 'I looked at the image.';
    case 'captions':
      return "I read the captions — I did NOT see the video, so I don't know what was shown on screen.";
    case 'youtube-text':
      return 'I only had the title and description — I neither watched nor heard it.';
    case 'article':
      return 'I read the article.';
    case 'document':
      return 'I read the document.';
    case 'text':
      return 'I read the text as given.';
    default:
      return 'I looked at what was provided.';
  }
}

// ---------- preparing content for a model to actually examine ----------

/**
 * Builds the `media` array for a model that can genuinely watch/see the
 * source, plus a `cleanup()` for anything uploaded.
 *
 * Two routes, cheapest first:
 *   INLINE   — small images (and PDFs, on an adapter rich enough) ride along
 *              as base64 in the request itself. No upload, works everywhere.
 *   UPLOADED — video, audio and anything over the inline cap go through
 *              `adapter.uploadFile`, which today only Gemini implements.
 *
 * `entry` is the model that will receive it — the uploaded route pins the
 * attachment to that one model's API key, so callers must not let the usual
 * fallback chain hand a dead URI to a different provider (askModel's
 * `only: true`).
 *
 * Returns { ok: true, media, cleanup } or { ok: false, error }.
 */
export async function prepareFor(entry, source) {
  const caps = getCapabilities(entry?.adapter);
  const adapter = getAdapter(entry.adapter);

  if (source.kind === 'youtube') {
    if (!caps.video) return { ok: false, error: 'That model cannot watch video.' };
    return {
      ok: true,
      media: [{ kind: 'video', mimeType: 'video/*', uri: source.url }],
      cleanup: async () => {},
    };
  }

  if (source.kind === 'file') {
    const kind = fileKind(source.filePath);
    const mimeType = mimeTypeFor(source.filePath);
    const isPdf = mimeType === 'application/pdf';

    if (kind === 'video' && !caps.video) return { ok: false, error: 'That model cannot watch video.' };
    if (kind === 'audio' && !caps.audio) return { ok: false, error: 'That model cannot listen to audio.' };
    if (kind === 'image' && !caps.vision) return { ok: false, error: 'That model cannot see images.' };
    if (isPdf && !caps.video) return { ok: false, error: 'That model cannot read a PDF directly.' };

    if (!fs.existsSync(source.filePath)) {
      return { ok: false, error: "I can't find that file — check the path is right." };
    }

    if (kind === 'image' || isPdf) {
      const inline = inlineAttachment(source.filePath, mimeType);
      if (inline.ok) return inline;
      if (inline.error !== 'too_big') return inline;
      // Over the inline cap — fall through to the upload route below.
    }

    if (typeof adapter.uploadFile !== 'function') {
      return {
        ok: false,
        error:
          kind === 'image' || isPdf
            ? 'That file is too big to send directly, and this model\'s provider has nowhere to upload it. A Gemini model can handle files this size.'
            : `Watching ${kind === 'audio' ? 'audio' : 'video'} needs a model that can take file uploads — Gemini can.`,
      };
    }

    try {
      const uploaded = await adapter.uploadFile(entry, source.filePath, mimeType);
      return {
        ok: true,
        media: [{ kind: kind === 'image' ? 'image' : 'video', mimeType: uploaded.mimeType, uri: uploaded.uri }],
        cleanup: uploaded.cleanup,
      };
    } catch (err) {
      return { ok: false, error: err?.message || "That file couldn't be uploaded." };
    }
  }

  return { ok: false, error: 'That kind of source is read as text, not watched.' };
}

File: backend/jarvis/content/intake.py
"""Working out what something IS, for free.

`identify()` makes NO model call, ever. That is the load-bearing part of this
package: it answers "what is this" from cheap metadata — YouTube's public oEmbed
endpoint, a page's own title, a file's size on disk — which is what lets Jarvis
say "that's a 40-minute video about X, what do you want from it?" and then
genuinely wait, rather than reading the thing first and asking afterwards.

The other half is `prepare_for()`: putting a source in front of a model that can
actually watch or see it, cheapest route first.
"""

from __future__ import annotations

import json
import logging
import re
from pathlib import Path
from typing import Any
from urllib.parse import parse_qs, urlparse

from ..media import file_kind, human_size, inline_attachment, mime_type_for, TooBig
from ..webtext import BROWSER_UA, fetch_article

logger = logging.getLogger(__name__)

#: YouTube serves a cookie-consent interstitial instead of the watch page to a
#: request carrying no consent cookie. These are what a browser would send.
YOUTUBE_HEADERS = {
    "User-Agent": BROWSER_UA,
    "Accept-Language": "en-US,en;q=0.9",
    "Cookie": "CONSENT=YES+cb.20210328-17-p0.en+FX+000; SOCS=CAI",
}

TRANSCRIPT_MAX_CHARS = 40000
_YOUTUBE_HOST = re.compile(r"^(www\.|m\.|music\.)?(youtube\.com|youtu\.be)$", re.I)
_YOUTUBE_PATH = re.compile(r"^/(shorts|embed|live|v)/([^/?#]+)")
_WINDOWS_PATH = re.compile(r"^[a-zA-Z]:[\\/]")


# --- what did the user actually hand over -------------------------------------

def youtube_video_id(url: str) -> str | None:
    try:
        parsed = urlparse(url)
    except ValueError:
        return None
    if not parsed.hostname or not _YOUTUBE_HOST.match(parsed.hostname):
        return None
    if parsed.hostname.lower().endswith("youtu.be"):
        return parsed.path.lstrip("/").split("/")[0] or None
    query = parse_qs(parsed.query or "")
    if query.get("v"):
        return query["v"][0]
    match = _YOUTUBE_PATH.match(parsed.path or "")
    return match.group(2) if match else None


def classify_source(text: str) -> dict[str, Any]:
    """A YouTube link, a web page, a file path, or plain text.

    Deliberately forgiving: it is fed by a text box AND a voice transcript, so it
    has to cope with a spoken path, a pasted link with tracking junk on it, or a
    wall of text with no marker at all. Purely syntactic — no network, no
    filesystem check, so it is safe to call on anything.
    """
    raw = (text or "").strip()
    if not raw:
        return {"kind": "text", "text": ""}

    video_id = youtube_video_id(raw)
    if video_id:
        return {"kind": "youtube", "url": raw, "videoId": video_id}

    if re.match(r"^https?://", raw, re.I):
        parsed = urlparse(raw)
        if parsed.netloc:
            return {"kind": "url", "url": raw}

    # "Copy as path" produces a quoted Windows path; a POSIX path is only
    # treated as one when it actually exists, since "/usr/bin is where..." is
    # far more likely to be a sentence than a file.
    unquoted = raw.strip('"').strip()
    if _WINDOWS_PATH.match(unquoted) or unquoted.startswith("\\\\"):
        return {"kind": "file", "filePath": unquoted}
    if unquoted.startswith("/") and "\n" not in unquoted:
        try:
            if Path(unquoted).is_file():
                return {"kind": "file", "filePath": unquoted}
        except OSError:
            pass

    return {"kind": "text", "text": raw}


# --- YouTube, without watching it ---------------------------------------------

def _oembed(url: str) -> dict[str, Any] | None:
    """Title and channel from YouTube's public oEmbed endpoint — no key, no
    scraping. The cheapest possible "what is this" for a video."""
    import httpx

    try:
        with httpx.Client(timeout=15.0, headers={"User-Agent": BROWSER_UA}) as client:
            response = client.get("https://www.youtube.com/oembed",
                                  params={"url": url, "format": "json"})
            if response.status_code >= 400:
                return None
            data = response.json()
    except Exception:  # noqa: BLE001
        return None
    return {"title": data.get("title"), "author": data.get("author_name"),
            "thumbnail": data.get("thumbnail_url")}


def _json_array_after(source: str, key: str) -> str | None:
    """The JSON array following `"key":` in a blob of page source.

    A regex genuinely cannot do this — the entries contain nested arrays, so any
    non-greedy match truncates mid-structure. Brackets are counted, respecting
    strings and escapes.
    """
    at = source.find(f'"{key}":')
    if at == -1:
        return None
    start = source.find("[", at)
    if start == -1:
        return None

    depth = 0
    in_string = False
    escaped = False
    for index in range(start, len(source)):
        char = source[index]
        if escaped:
            escaped = False
            continue
        if char == "\\":
            escaped = True
            continue
        if char == '"':
            in_string = not in_string
        if in_string:
            continue
        if char == "[":
            depth += 1
        elif char == "]":
            depth -= 1
            if depth == 0:
                return source[start:index + 1]
    return None


def _captions(xml: str) -> str:
    from ..webtext import to_text

    lines = [to_text(body).strip()
             for body in re.findall(r"(?s)<text[^>]*>(.*?)</text>", xml)]
    return " ".join(line for line in lines if line)


def fetch_youtube_text(url: str) -> dict[str, Any]:
    """Everything about a video that can be had without watching it.

    **Expect no transcript in practice.** YouTube serves empty caption bodies to
    anything without a session token, and this build is not going to pretend
    otherwise. Title and description come from sturdier sources and do work.
    That gap is exactly why watching properly matters when it is available, and
    why the caller must say which of the two it actually had.
    """
    import httpx

    meta = _oembed(url) or {}
    result: dict[str, Any] = {"ok": True, "title": meta.get("title"),
                              "author": meta.get("author"),
                              "thumbnail": meta.get("thumbnail"),
                              "description": None, "transcript": None}

    markup = ""
    try:
        with httpx.Client(timeout=20.0, follow_redirects=True,
                          headers=YOUTUBE_HEADERS) as client:
            response = client.get(url, params={"hl": "en", "gl": "US"})
            if response.status_code < 400:
                markup = response.text
    except Exception:  # noqa: BLE001 — oEmbed alone still says what the video is
        markup = ""

    if not markup:
        if not result["title"]:
            return {"ok": False, "error": "Couldn't reach that video."}
        return result

    described = re.search(r'"shortDescription":"((?:\\.|[^"\\])*)"', markup)
    if described:
        try:
            result["description"] = json.loads(f'"{described.group(1)}"')
        except ValueError:
            result["description"] = described.group(1)

    tracks_json = _json_array_after(markup, "captionTracks")
    if tracks_json:
        try:
            tracks = json.loads(tracks_json.replace("\\u0026", "&"))
            track = (next((t for t in tracks
                           if str(t.get("languageCode", "")).startswith("en")
                           and t.get("kind") != "asr"), None)
                     or next((t for t in tracks
                              if str(t.get("languageCode", "")).startswith("en")), None)
                     or (tracks[0] if tracks else None))
            if track and track.get("baseUrl"):
                with httpx.Client(timeout=20.0, headers=YOUTUBE_HEADERS) as client:
                    captions = client.get(track["baseUrl"])
                if captions.status_code < 400:
                    transcript = _captions(captions.text)
                    if transcript:
                        result["transcript"] = transcript[:TRANSCRIPT_MAX_CHARS]
        except Exception:  # noqa: BLE001 — no captions is the common case
            logger.debug("no usable captions for %s", url, exc_info=True)

    return result


# --- the free glance ----------------------------------------------------------

def identify(source: dict[str, Any]) -> dict[str, Any]:
    """What this content IS. No model call, ever — see this module's docstring.

    For a URL this does fetch the page (a title needs it) and keeps the text as
    `cachedText`, so one fetch serves both "what is this" and the later reading
    rather than paying for two.
    """
    kind = source.get("kind")

    if kind == "youtube":
        meta = _oembed(source["url"]) or {}
        return {"title": meta.get("title") or f"YouTube video ({source.get('videoId')})",
                "kind": "video",
                "detail": f"on {meta['author']}" if meta.get("author") else None,
                "thumbnail": meta.get("thumbnail")}

    if kind == "url":
        article = fetch_article(source["url"])
        if not article.ok:
            return {"title": source["url"], "kind": "article", "detail": None,
                    "unreachable": article.error}
        return {"title": article.title or source["url"], "kind": "article",
                "detail": f"about {max(1, round(len(article.text) / 1000))}k characters",
                "cachedText": article.text}

    if kind == "file":
        path = Path(source["filePath"])
        try:
            size = path.stat().st_size
        except OSError:
            return {"title": path.name, "kind": file_kind(path), "detail": None,
                    "unreachable": "I can't find that file."}
        return {"title": path.name, "kind": file_kind(path), "detail": human_size(size)}

    text = str(source.get("text") or "")
    first_line = text.splitlines()[0].strip() if text.strip() else ""
    title = (f"{first_line[:60]}…" if len(first_line) > 60 else first_line) or "Pasted text"
    return {"title": title, "kind": "text", "detail": f"{len(text)} characters",
            "cachedText": text}


_KIND_WORDS = {"video": "a video", "article": "a web article", "document": "a document",
               "image": "an image", "audio": "an audio file", "text": "some pasted text",
               "unknown": "a file"}


def describe(identity: dict[str, Any]) -> str:
    """Plain-language "what this is", for Jarvis to say straight after the free
    glance — before it has read or watched anything."""
    bits = [_KIND_WORDS.get(identity.get("kind"), "something")]
    if identity.get("title"):
        bits.append(f"— \"{identity['title']}\"")
    if identity.get("detail"):
        bits.append(f"({identity['detail']})")
    return " ".join(bits)


def intake_description(kind: str | None) -> str:
    """How content was actually taken in, said honestly every time, so an answer
    never sounds better-informed than it is."""
    return {
        "video": "I watched the video itself, including what was on screen.",
        "audio": "I listened to the audio.",
        "image": "I looked at the image.",
        "captions": ("I read the captions — I did NOT see the video, so I don't know what "
                     "was shown on screen."),
        "youtube-text": "I only had the title and description — I neither watched nor heard it.",
        "article": "I read the article.",
        "document": "I read the document.",
        "text": "I read the text as given.",
    }.get(kind or "", "I looked at what was provided.")


# --- putting it in front of a model -------------------------------------------

class CannotPrepare(RuntimeError):
    """This model cannot take this source. The message is for the user."""


def prepare_for(entry: dict[str, Any], source: dict[str, Any]) -> dict[str, Any]:
    """Media for a model that can genuinely watch or see this, plus a cleanup.

    Two routes, cheapest first: INLINE base64 (no upload, works on every
    adapter) for small images and PDFs, and UPLOADED for video, audio, and
    anything over the inline cap. The uploaded route pins the file to ONE
    model's API key, which is why its caller must disable fallback.
    """
    from ..adapters import get_adapter, get_capabilities

    caps = get_capabilities(entry.get("adapter"))
    adapter = get_adapter(entry.get("adapter"))

    if source.get("kind") == "youtube":
        if not caps.get("video"):
            raise CannotPrepare("That model cannot watch video.")
        return {"media": [{"kind": "video", "mimeType": "video/*", "uri": source["url"]}],
                "cleanup": None}

    if source.get("kind") != "file":
        raise CannotPrepare("That kind of source is read as text, not watched.")

    path = Path(source["filePath"])
    kind = file_kind(path)
    mime_type = mime_type_for(path)
    is_pdf = mime_type == "application/pdf"

    if kind == "video" and not caps.get("video"):
        raise CannotPrepare("That model cannot watch video.")
    if kind == "audio" and not caps.get("audio"):
        raise CannotPrepare("That model cannot listen to audio.")
    if kind == "image" and not caps.get("vision"):
        raise CannotPrepare("That model cannot see images.")
    if is_pdf and not caps.get("video"):
        raise CannotPrepare("That model cannot read a PDF directly.")
    if not path.exists():
        raise CannotPrepare("I can't find that file — check the path is right.")

    if kind == "image" or is_pdf:
        try:
            return {"media": inline_attachment(path, mime_type), "cleanup": None}
        except TooBig:
            pass                         # over the cap: fall through to uploading
        except OSError as err:
            raise CannotPrepare("I couldn't read that file.") from err

    upload = getattr(adapter, "upload_file", None)
    if not callable(upload):
        raise CannotPrepare(
            "That file is too big to send directly, and this model's provider has nowhere "
            "to upload it. A Gemini model can handle files this size."
            if kind == "image" or is_pdf else
            f"Taking in {'audio' if kind == 'audio' else 'video'} needs a model that can "
            "accept file uploads — Gemini can.")

    uploaded = upload(entry, str(path), mime_type)
    return {"media": [{"kind": "image" if kind == "image" else "video",
                       "mimeType": uploaded.get("mimeType", mime_type),
                       "uri": uploaded["uri"]}],
            "cleanup": uploaded.get("cleanup")}
